design/design-doc-composition-functions.md

```diff
@@ -175,604 +183,831 @@ metadata:
 
-Support for `type: Chroot` functions could be added shortly after initial
-support for `type: Container` functions are released if `type: Container` proves
-to be insufficiently compatible (e.g. for clusters running gVisor, or that
-require seccomp be enabled).
+These metrics pertain to Kubernetes controllers, and will be insufficient when a
+significant portion of Composition logic is 'outsourced' to Functions. It will
+be important to establish a pattern for instrumenting Functions (e.g. as part of
+the Function SDKs). Request tracing in particular is likely to be useful in
+order to debug slow or failing Functions in a pipeline.
 
-### Using Kubernetes to Run Containerized Functions
+Metrics and tracing for Functions must be implemented before the feature becomes
+generally available (GA).
 
-Asking Kubernetes to run a container pipeline is less straightforward than you
-might think. Crossplane could schedule a `Pod` for each XR reconcile, or create
-a `CronJob` to do so regularly. Another option could be to connect directly to a
-Kubelet. This approach would enjoy all the advantages of the existing Kubelet
-machinery (pull secrets, caching, etc) but incurs overhead in other areas, for
-example:
+#### Caching
 
-* Every reconcile requires a pod to be scheduled, which may potentially block on
-  node scale-up, etc.
-* stdin and stdout must be streamed via the API server, for example by using the
-  [`/attach` subresource][attach].
-* Running containers in order requires either (ab)using init containers or
-  injecting a middleware binary that blocks container starts to ensure they run
-  in order (similar to Argo Workflow's '[emissary]' executor):
+The gRPC API proposed by this design accomodates caching, but building caching
+support is not in scope for beta. This will likely be a requirement for GA.
 
-> The emissary works by replacing the container's command with its own command.
-> This allows that command to capture stdout, the exit code, and easily
-> terminate your process. The emissary can also delay the start of your process.
+#### Runtime Configuration
 
-You can see some of the many options Argo Workflows explored to address these
-issues before landing on `emissary` in their list of
-[deprecated executors][argo-deprecated-executors].
+This design proposes Functions be long-running processes, installed by the
+package manager. Deploying Functions as Kubernetes Deployments (with Services,
+Service Accounts, etc) will no doubt necessitate something like a Provider's
+ControllerConfig type.
 
-### Using KRM Function Spec Compliant Functions
+We've identified that we [don't want to proceed][controllerconfig-deprecation]
+with ControllerConfig, but don't yet have a suitable alternative. Rather than
+propagating a ControllerConfig-like pattern I propose we prioritize finding an
+alternative. I intend to open a separate, simultaneous design to address this
+since it will affect Provider packages as well as Functions.
```

Comment:
The cliff-notes version of my impending proposal, which I hope will be a "good enough" alternative to a PRI (https://github.com/crossplane/crossplane/issues/2671):

Add a core Crossplane CLI flag like `--package-runtime` with two values - `Kubernetes` (the default) and `External`. When running in "Kubernetes" mode the package manager does what it does today - assumes it's running on a Kubernetes cluster and makes Deployments, Services etc to run Providers and Functions. When running in "External" mode the package manager revision controllers are disabled. e.g. When you create a `Provider` the package manager creates a `ProviderRevision` but doesn't reconcile it. It expects some external system (such as a controller running alongside Crossplane) to figure out what to do with the package revision.

The rationale here is that AFAIK today 99% of folks are running Crossplane in a Kubernetes cluster (or somewhere that quacks like a Kubernetes cluster - e.g. vcluster). It's really only advanced folks (i.e. vendors like Upbound) who might want to deploy extensions in a completely different way, so letting them write a controller to do that seems reasonable.

To solve the `ControllerConfig`-is-just-a-template-for-Deployment problem I propose we lean in and actually make it a template for `Deployment` (and friends - e.g. `Service`). I imagine this would look similar to a Deployment's pod spec template. I'd remove most of the hardcoded templating from the package manager codebase and push it into a `default` config (like the `default` `ProviderConfig` pattern). Given Functions aren't a controller I'd call this a `RuntimeConfig`.

CC @hasheddan